central/auth/m2m/set.go

```diff
@@ -0,0 +1,117 @@
+	UpsertTokenExchanger(exchanger TokenExchanger, issuer string) error
+	RemoveTokenExchanger(issuer string) error
+	GetTokenExchanger(issuer string) (TokenExchanger, bool)
+}
+
+// TokenExchangerFactory factory for creating a new token exchanger.
+type TokenExchangerFactory = func(ctx context.Context, config *storage.AuthMachineToMachineConfig, roleDS roleDataStore.DataStore, issuerFactory tokens.IssuerFactory) (TokenExchanger, error)
+
+// TokenExchangerSetSingleton creates a singleton holding all token exchangers for auth machine to machine configs.
+func TokenExchangerSetSingleton(roleDS roleDataStore.DataStore, issuerFactory tokens.IssuerFactory) TokenExchangerSet {
+	once.Do(func() {
+		t = &tokenExchangerSet{
+			tokenExchangers:       map[string]TokenExchanger{},
+			roleDS:                roleDS,
+			issuerFactory:         issuerFactory,
+			tokenExchangerFactory: newTokenExchanger,
+		}
+	})
+	return t
+}
+
+// TokenExchangerSetForTesting creates a token set for testing purposes.
+func TokenExchangerSetForTesting(_ *testing.T, roleDS roleDataStore.DataStore, issuerFactory tokens.IssuerFactory,
+	tokenExchangerFactory TokenExchangerFactory) TokenExchangerSet {
+	return &tokenExchangerSet{
+		tokenExchangers:       map[string]TokenExchanger{},
+		roleDS:                roleDS,
+		issuerFactory:         issuerFactory,
+		tokenExchangerFactory: tokenExchangerFactory,
+	}
+}
+
+type tokenExchangerSet struct {
+	tokenExchangers       map[string]TokenExchanger
+	roleDS                roleDataStore.DataStore
+	issuerFactory         tokens.IssuerFactory
+	tokenExchangerFactory TokenExchangerFactory
+}
+
+// NewTokenExchangerFromConfig creates a TokenExchanger from the given config.
+// It will make sure the TokenExchanger is registered as a tokens.Source.
+// Note that during creation of the TokenExchanger, an external HTTP request is done to the OIDC issuer
+// to retrieve additional metadata for token validation (i.e. JWKS metadata).
+func (t *tokenExchangerSet) NewTokenExchangerFromConfig(ctx context.Context, config *storage.AuthMachineToMachineConfig) (TokenExchanger, error) {
+	exchanger, exists := t.tokenExchangers[config.GetIssuer()]
+	if exists {
+		// Need to unregister the source temporarily, otherwise we receive an error on creation that the
+		// source is already registered.
+		if err := t.issuerFactory.UnregisterSource(exchanger.Provider()); err != nil {
+			return nil, pkgErrors.Wrapf(err, "unregistering source for config %s", config.GetId())
+		}
+	}
+
+	tokenExchanger, err := t.tokenExchangerFactory(ctx, config, t.roleDS, t.issuerFactory)
+	if err != nil {
+		return nil, pkgErrors.Wrapf(err, "creating token exchanger for config %s", config.GetId())
+	}
+
+	return tokenExchanger, nil
+}
+
+// GetTokenExchanger retrieves a TokenExchanger based on the issuer.
+func (t *tokenExchangerSet) GetTokenExchanger(issuer string) (TokenExchanger, bool) {
+	tokenExchanger, exists := t.tokenExchangers[issuer]
```

Comment:
I don't necessarily want us to be extremely smart about it, but it seems that the issuer could be provided with or without a trailing slash, e.g. `https://token.actions.githubusercontent.com` vs `https://token.actions.githubusercontent.com/`, which means the same.

What do you think about normalizing the trailing slash both in the config and in the issuer url we received in the token so that matching in the set can work no matter how it's [mis-]configured?

Obviously, there can be cases with multiple slashes used in the middle of the path which again create different literal strings for the same actual resource (`https://my.idp.blah/foo//bar///baz/` vs `https://my.idp.blah/foo/bar/baz/`). I don't suggest to address these unless there's a ready-made battle-tested function in StackRox for such normalization.

I think, this `tokenExchangerSet` might be the right place to do the normalization because it needs it for matching m2m config with incoming tokens. Which means the key in the map can be normalized whereas the issuer url provided in the config may remain as the user provided it.

Reply:
See [https://github.com/stackrox/stackrox/pull/8145#discussion_r1370079760]

Reply:
I don't think we need normalization here, why should we? The issuer is used by the oidc library for verification purposes, for us internally it's only used for a key in the token exchanger map.

I don't see a benefit of doing this normalization at the current time.

Reply:
Let's not do the normalization since higher-level RFC also suggests us not to. Resolving.